include/sycl/algorithm/buffer_algorithms.hpp

```diff
@@ -0,0 +1,541 @@
+  //number of elements manipulated by each work_group (except the last one)
+  size_t size_per_work_group = size_per_work_item * nb_work_item;
+  //std::cout << "size_per_work_group=\t" << size_per_work_group << std::endl;
+
+  nb_work_group = max(static_cast<size_t>(1),
+                      up_rounded_division(size, size_per_work_group));
+  //std::cout << "nb_work_group=\t" << nb_work_group
+  //          << " (updated)" << std::endl;
+  assert(nb_work_group >= 1);
+
+  assert(size_per_work_group * (nb_work_group - 1) < size);
+  assert(size_per_work_group * nb_work_group >= size);
+  /* number of elements manipulated by the last work_group
+   * n.b. if the value is 0, the last work_group is regular
+   */
+  /*size_t size_last_work_group = size % size_per_work_group;*/
+  //std::cout << "size_last_work_group=" << size_last_work_group << std::endl;
+
+  /*size_t size_per_work_item_last = up_rounded_division(size_last_work_group,
+                                                       nb_work_item);*/
+
+  return sycl_algorithm_descriptor {
+    size,
+    size_per_work_group,
+    size_per_work_item,
+    nb_work_group,
+    nb_work_item };
+}
+
+/*
+ * MapReduce Algorithm applied on a buffer
+ *
+ * with map/reduce functions typed as follow
+ * Map    : A -> B
+ * Reduce : B -> B -> B
+ */
+
+template <typename ExecutionPolicy,
+          typename A,
+          typename B,
+          typename Reduce,
+          typename Map>
+B buffer_mapreduce(ExecutionPolicy &snp,
+                   cl::sycl::queue q,
+                   cl::sycl::buffer<A, 1> input_buff,
+                   B init, //map is not applied on init
+                   sycl_algorithm_descriptor d,
+                   Map map,
+                   Reduce reduce) {
+
+  /*
+   * 'map' is not applied on init
+   * 'map': A -> B
+   * 'reduce': B * B -> B
+   */
+
+  if ((d.nb_work_item == 0) || (d.nb_work_group == 0)) {
+    auto read_input = input_buff.template get_access
+      <cl::sycl::access::mode::read, cl::sycl::access::target::host_buffer>();
+    B acc = init;
+    for (size_t pos = 0; pos < d.size; pos++)
+      acc = reduce(acc, map(pos, read_input[pos]));
+
+    return acc;
+  }
+
+  using std::min;
+  using std::max;
+
+  cl::sycl::buffer<B, 1> output_buff { cl::sycl::range<1> { d.nb_work_group } };
+
+  q.submit([&] (cl::sycl::handler &cgh) {
+    cl::sycl::nd_range<1> rng
+      { cl::sycl::range<1>{ d.nb_work_group * d.nb_work_item },
+        cl::sycl::range<1>{ d.nb_work_item } };
+    auto input = input_buff.template get_access
+      <cl::sycl::access::mode::read>(cgh);
+    auto output = output_buff.template get_access
+      <cl::sycl::access::mode::write>(cgh);
+    cl::sycl::accessor<B, 1, cl::sycl::access::mode::read_write,
+                       cl::sycl::access::target::local>
+      sum { cl::sycl::range<1>(d.nb_work_item), cgh };
+    cgh.parallel_for_work_group<class wg>(rng, [=](cl::sycl::group<1> grp) {
```

Comment:
The constructor on the specification takes a range, not an nd_range, and there is no implicit conversion among them (This fails to compile in computecpp)